lasagne/layers/pool.py

```diff
@@ -755,3 +756,161 @@ def get_output_shape_for(self, input_shape):
+    level :math:`i`, and we use a list of the :math:`n_i`'s as a
+    parameter for SPP-Layer.
+    The length of this list is the level of the spatial pyramid.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or tuple
+        The layer feeding into this layer, or the expected input shape.
+
+    pool_dims : list of integers
+        The list of :math:`n_i`'s that define the output dimension of each
+        pooling level :math:`i`. The length of pool_dims is the level of
+        the spatial pyramid.
+
+    mode : string
+        Pooling mode, one of 'max', 'average_inc_pad', 'average_exc_pad' or
+        'kaiming'. Defaults to 'max'.
+
+    **kwargs
+        Any additional keyword arguments are passed to the :class:`Layer`
+        superclass.
+
+    Notes
+    -----
+    This layer should be inserted between the convolutional part of a
+    DNN and its dense part. Convolutions can be used for
+    arbitrary input dimensions, but the size of their output will
+    depend on their input dimensions. Connecting the output of the
+    convolutional to the dense part then usually demands us to fix
+    the dimensions of the network's InputLayer.
+    The spatial pyramid pooling layer, however, allows us to leave the
+    network input dimensions arbitrary. The advantage over a global
+    pooling layer is the added robustness against object deformations
+    due to the pooling on different scales.
+
+    Using mode='kaiming' implements the layer as defined in the appendix A
+    of [1], which uses non-uniform strides and window sizes. This mode works
+    on any input size, but it is slower.
+
+    References
+    ----------
+    .. [1] He, Kaiming et al (2015):
+           Spatial Pyramid Pooling in Deep Convolutional Networks
+           for Visual Recognition.
+           http://arxiv.org/pdf/1406.4729.pdf.
+    """
+    def __init__(self, incoming, pool_dims=[4, 2, 1], mode='max', **kwargs):
+            super(SpatialPyramidPoolingLayer, self).__init__(incoming,
+                                                             **kwargs)
+            if len(self.input_shape) != 4:
+                raise ValueError("Tried to create a SPP layer with "
+                                 "input shape %r. Expected 4 input dimensions "
+                                 "(batchsize, channels, 2 spatial dimensions)."
+                                 % (self.input_shape,))
+
+            if mode != 'kaiming':
+                # Check if the running theano version supports symbolic
+                # variables as arguments for pool_2d. This is required
+                # unless using mode='kaiming'
+                try:
+                    pool_2d(T.tensor4(),
+                            ws=T.ivector(),
+                            stride=T.ivector(),
+                            ignore_border=True,
+                            pad=None)
+                except (ValueError, TypeError):
```

Comment:
I had put it there while I was performing the check with T.signal.pool.pool_2d. A TypeError was raised if the theano version had the old parameter names for pool_2d. Now that we're using lasagne.pool.pool_2d (which does this check), this is no longer necessary. I have removed it in the new commit